internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -851,7 +851,32 @@ func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			}}}
 		}
 
-		bmh, err := r.findBMH(ctx, agent)
+		bmh, err := r.findBMHByAgent(ctx, agent)
+		if bmh == nil || err != nil {
+			return []reconcile.Request{}
+		}
+
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: bmh.Namespace,
+			Name:      bmh.Name,
+		}}}
+	}
+
+	mapInfraEnvToBMH := func(a client.Object) []reconcile.Request {
+		ctx := context.Background()
+		infraEnv := &aiv1beta1.InfraEnv{}
+
+		if err := r.Get(ctx, types.NamespacedName{Name: a.GetName(), Namespace: a.GetNamespace()}, infraEnv); err != nil {
+			return []reconcile.Request{}
+		}
+
+		// Don't queue any reconcile if the InfraEnv
+		// doesn't have the ISODownloadURL set yet.
+		if infraEnv.Status.ISODownloadURL == "" {
+			return []reconcile.Request{}
+		}
+
+		bmh, err := r.findBMHByInfraEnv(ctx, infraEnv)
```

Comment:
This code is finding the first BMH that matches and immediately returning a request just for it. Instead it should continue iterating BMHs, find all of them that match the InfraEnv, and return requests for all of them.

(trying a new review because just replying did not un-resolve the thread)

Reply:
Oooooh, I see what you mean. I was confused because you commented here, instead of commenting in the `findBMHByInfraEnv` function which, as you correctly suggest, should be returning the list of BMHs associated with a single `InfraEnv`.

Let me fix this, thanks for catching it and explaining it again :pray: